[SYCL] Introduce sycl complex marray specialization

jle-quel opened the pull request:
Following the introduction of `sycl::ext::oneapi::experimental::complex` (https://github.com/intel/llvm/pull/8068), which adds the support of complex for SYCL, this PR allows for adding new complex math features.

The specialization of `sycl::marray` to add support for storing complex numbers in arrays is proposed as well as overloading the existing math functions to support complex marrays and adding member functions to simplify accessing, setting marray values, and constructing complex marrays.

jle-quel commented:
The latest commits are the propagation of this PR https://github.com/intel/llvm-test-suite/pull/1661 which was on the recently archived llvm-test-suite repository.

jle-quel commented on sycl/include/sycl/ext/oneapi/experimental/sycl_complex.hpp:
```diff
@@ -61,9 +61,7 @@ template <class _Tp> struct __numeric_type {
-template <> struct __numeric_type<void> {
-  static const bool value = true;
-};
+template <> struct __numeric_type<void> { static const bool value = true; };
```
I agree with you, and the reason for using static const bool is due to inheritance.
The sycl::complex interface is based on the `libcxx/complex` and `libcxx/complex` use the `promote.h`, which defines these `value` as `static const bool`.

I could within another PR modifies that for the sycl::complex

jle-quel commented on sycl/include/sycl/ext/oneapi/experimental/sycl_complex.hpp:
```diff
@@ -983,6 +981,426 @@ tan(const complex<_Tp> &__x) {
+
+  explicit constexpr marray(const ComplexDataT &arg) {
+    for (size_t i = 0; i < NumElements; ++i) {
+      MData[i] = arg;
```
I will propagate the changes you made with your PR https://github.com/intel/llvm/commit/ba0ea790374a0194d022ba6a633ee679c749acee in this PR

jle-quel commented:
Is there anything else you wanted me to update regarding this PR @steffenlarsen ?

jle-quel commented on sycl/include/sycl/ext/oneapi/experimental/sycl_complex.hpp:
```diff
@@ -983,6 +983,480 @@ tan(const complex<_Tp> &__x) {
+  // Alias for shortening the marray arguments to array converter.
+  template <typename ComplexDataT, typename... ArgTN>
+  using MArrayArgArrayCreator =
+      detail::ArrayCreator<ComplexDataT, FlattenMArrayArg, ArgTN...>;
```
Absolutely. Would bringing these helpers outside of the `marray` class and inside the `detail` namespace of the same `sycl/include/sycl/marray.hpp` file, would work for you?

steffenlarsen replied:
Sure!

jle-quel replied:
What's the reason for this FIXME you added to your latest commit?

```
// FIXME: Other marray specializations needs to be a friend to access MData.
// If the subscript operator is made constexpr this can be removed.
template <typename Type_, std::size_t NumElements_> friend class marray;
```

I'm asking because, to move the helper traits outside of the class `marray`, and inside the namespace ```sycl::detail``` (since one of the traits is accessing directly MData), I had to use the subscript operator on the `marray` and make it a `constexpr`.

steffenlarsen replied:
You are right, making the subscript operator `constexpr` solves the issue, but that is not actually in line with the specification. I opened an issue about this here: https://github.com/KhronosGroup/SYCL-Docs/issues/367.

As a work-around, the friend here lets us skip the subscript operator and get the data directly. Maybe an option is to move the implementation into a helper struct that is in turn a friend of `marray` until this is resolved?

jle-quel replied:
That could be a temporary solution, yes.
I'll give some time to your issue to "grow", and if `constexpr` is soon added to the spec, I'll keep the PR as is, otherwise, I'll introduce a FIXME: temporary solution.

steffenlarsen replied:
The issue is from February, so I don't think it's worth holding our breaths for a resolution soon.

jle-quel replied:
I hesitated between a friend struct and adding a "temporary" getter.

The problem with temporary getter is that this public getter might be used by and will be "harder" to remove if codebases depend on it.

Encapsulating the helper functions into a helper struct without renaming symbols and other "big" modifications looked the easiest to transit when the subscript operators will be `constexpr`.

aelovikov-intel commented on sycl/include/sycl/ext/oneapi/experimental/sycl_complex.hpp:
```diff
@@ -1003,6 +1003,448 @@ constexpr complex<float> operator""if(unsigned long long __im) {
+// Special definition as polar requires default argument
+
+template <typename T, std::size_t NumElements,
+          typename = std::enable_if<is_genfloat<T>::value>>
+_SYCL_EXT_CPLX_INLINE_VISIBILITY
+    sycl::marray<sycl::ext::oneapi::experimental::complex<T>, NumElements>
```
I think I'd prefer using `std::enable_if` on the return type.

aelovikov-intel replied:
> Is it just for this case?

No, generally.

> Why do you prefer it?

Because it doesn't require an artificial parameter that could technically be supplied by the caller (even though it's not supposed to be done).

aelovikov-intel commented on sycl/test-e2e/Complex/sycl_complex_marray_getters_test.cpp:
```diff
@@ -0,0 +1,92 @@
+// RUN: %{build} -fsycl-device-code-split=per_kernel -o %t.out
```
Is `-fsycl-device-code-split=per_kernel` necessary? If so, please add a comment why, otherwise remove it.

jle-quel commented:
Latest updates:
- SYCL end-to-end test suites for Windows were failing due to special test cases that can't be checked for equality with other complex implementations.
- This same test suite for Windows, generated some ambiguity errors. (Window's complex implementation for windows is unspecified for other types than `float`, `double`, `ldouble`).

Both failures have been fixed.

Finally, I believe that I addressed all your feedback @aelovikov-intel @steffenlarsen.
If that's not the case, let me know; otherwise, we can move forward with this PR.

aelovikov-intel commented on sycl/include/sycl/ext/oneapi/experimental/sycl_complex.hpp:
```diff
@@ -982,6 +982,451 @@ tan(const complex<_Tp> &__x) {
+  const_iterator end() const { return MData + NumElements; }
+
+  // OP is: +, -, *, /
+#define OP(op)                                                                 \
```
Please use a name that has less probability to conflict with user's macros.

jle-quel commented:
Gentle ping to get some attention into this PR 👍 @aelovikov-intel @steffenlarsen

steffenlarsen reviewed (commented):
I just have one request. Am I right that this is blocked by https://github.com/intel/llvm/pull/8852?

jle-quel commented on sycl/include/sycl/ext/oneapi/experimental/sycl_complex.hpp:
```diff
@@ -1010,6 +1010,464 @@ __DPCPP_SYCL_EXTERNAL _SYCL_EXT_CPLX_INLINE_VISIBILITY
+using marray_data_t = typename detail::vec_helper<T>::RetType;
+
+template <typename T, std::size_t NumElements>
+class marray<sycl::ext::oneapi::experimental::complex<T>, NumElements> {
```
That sounds great! I actually mentioned and proposed that in a draft PR https://github.com/intel/llvm/pull/10854

Would it be better for the `complex` and `marray`, to live in a directory `complex` under `sycl/include/sycl/ext/oneapi/experimental/` where one file will be the `complex` header, one file will be the `marray` header and one would be the common details: helper functions, traits ... that are shared by the complex headers?

Otherwise for simplicity, I'll just split the two and include complex into marray.

What do you think?

steffenlarsen replied:
I am okay with both! The former may mean less inter-dependencies which may make it easier to make changes to in the future, so slight preference for that from my side. 😄

jle-quel commented:
> I just have one request. Am I right that this is blocked by #8852?

It's not really blocked, since there is no build dependency between the implementation and the proposal, however, yes they represent the same thing so they should be merged together.

steffenlarsen commented:
Typically we would want the extension approved and merged before we consider merging the implementation of it. As such https://github.com/intel/llvm/pull/8852 should be first, I would argue.

steffenlarsen commented on sycl/include/sycl/ext/oneapi/experimental/complex/detail/marray.hpp:
```diff
@@ -0,0 +1,316 @@
+
+  MARRAY_CPLX_OP(&=)
+  MARRAY_CPLX_OP(|=)
+  MARRAY_CPLX_OP(^=)
```
Why not have `%=` deleted here too?

jle-quel replied:
The reason for not having the `%=` operator being deleted with the `bitwise` assignment operator is because I arranged the operator by "type". The `modulo` operator not being a `bitwise`  operation felt like it had to be on its own.

- Arithmetic
- Arithmetic assignment
- Bitwise
- Bitwise assignment
- etc ...

steffenlarsen commented on sycl/include/sycl/ext/oneapi/experimental/complex/detail/marray.hpp:
```diff
@@ -0,0 +1,316 @@
+// MARRAY_CPLX_OP is: &, |, ^
+#define MARRAY_CPLX_OP(op)                                                     \
+  friend marray operator op(const marray &lhs, const marray &rhs) = delete;    \
+  friend marray operator op(const marray &lhs, const value_type &rhs) = delete;
```
Is there a reason we don't delete
```c++
friend marray operator op(const value_type &lhs, const marray &rhs)
```
like we do for % and bitshifts? Seems like we could just have a `MARRAY_CPLX_OP_DELETE` and just delete all of them in one... Or two if we do the same for op-assigments.

jle-quel replied:
Sure! As I said in my previous comment. This was mostly arranged by type as a personal preference.
I'm happy to have a generic macro which deletes all operators.

jle-quel replied:
Please correct me if I did not understand correctly your feedback. Still, the reason for not having a generic `macro` is because of the different arguments that all operators received.

We can't place all the operators in the same basket due to their arguments.

steffenlarsen commented on sycl/include/sycl/ext/oneapi/experimental/complex/detail/marray_math.hpp:
```diff
@@ -0,0 +1,191 @@
+} // namespace ext
+
+} // namespace _V1
+} // namespace sycl
```
@aelovikov-intel - With the builtin refactoring, can we reuse some of the new tooling for this?

aelovikov-intel replied:
I'd say the `NUM_ARGS` macro parameter approach looks promising, although that's more of an idiom than tooling. Having a `marray_map` helper would remove some code duplication as well.

jle-quel replied:
I've updated the builtins and explained the details in my last comment

steffenlarsen commented on sycl/include/sycl/marray.hpp:
```diff
@@ -39,43 +39,30 @@ template <typename T, typename... Ts> struct GetMArrayArgsSize<T, Ts...> {
+struct AllSuitableArgTypes
+    : std::conjunction<IsSuitableArgType<DataT, ArgTN>...> {};
 
+class HelperFlattenMArrayArg {
```
Nit: I would probably have named it `FlattenMArrayArgHelper`, but it's a very small preference.

steffenlarsen commented on sycl/include/sycl/marray.hpp:
```diff
@@ -85,20 +72,44 @@ template <typename Type, std::size_t NumElements> class marray {
+private:
+  value_type MData[NumElements];
+
+  /// FIXME: If the subscript operator is made constexpr these can be removed.
+  // Other marray specializations needs to be a friend to access MData.
```
Nit: I don't mind changing it so that `FIXME` is only the actionable part of the comment, but I would then prefer it to be the later part of the comment:
```suggestion
  // Other marray specializations needs to be a friend to access MData.
  // FIXME: If the subscript operator is made constexpr these can be removed.
```

steffenlarsen commented on sycl/include/sycl/marray.hpp:
```diff
@@ -85,20 +72,44 @@ template <typename Type, std::size_t NumElements> class marray {
   template <typename Type_, std::size_t NumElements_> friend class marray;
+  // detail::HelperFlattenMArrayArg::MArrayToArray needs to be a friend to
+  // access MData.
+  friend class detail::HelperFlattenMArrayArg;
```
I suspect this can have the same `FIXME` applied as above. Maybe the intention was to do that by moving it, but I would prefer we just re-mention it so whoever addresses this will know both can go.

steffenlarsen reviewed (commented):
Just a handful additional comments, then I think I'm happy with it.

jle-quel commented:
Have you seen this https://github.com/intel/llvm/pull/12786 @steffenlarsen?

steffenlarsen commented:
> Have you seen this #12786 @steffenlarsen?

I missed that! Thanks for pointing it out. 😄

jle-quel commented:
> > The last comment that needs to be resolved before this PR can get merged is the one at "[#8647 (comment)](https://github.com/intel/llvm/pull/8647#discussion_r1484301864)".
> > I'm not sure I fully understand what needs to be done. Could you please be more specific?
> 
> We refactored the builtins, so the comment was about making these new builtins similar in definition to the new structure. See https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/detail/builtins/math_functions.inc and https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/detail/builtins/helper_macros.hpp.

I've updated the implementation of the marray builtin to follow the new structure of the builtins. However, it is a bit more convoluted since we are merging a common interface but with different arguments and logic within that function.

I've added some comments to help understand the macros and how to create a new builtin if needed in the future.

Let me know what you think @steffenlarsen
